risc0/zkvm/src/groth16.rs

```diff
@@ -0,0 +1,802 @@
+        let mut pvk_bytes = Vec::new();
+        let public_key_verification = Self::pvk()?;
+        public_key_verification
+            .serialize_uncompressed(&mut pvk_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let mut proof_bytes = Vec::new();
+        let proof = Proof::<Bn254> {
+            a: g1_from_bytes(&groth16_seal.a)?,
+            b: g2_from_bytes(&groth16_seal.b)?,
+            c: g1_from_bytes(&groth16_seal.c)?,
+        };
+        proof
+            .serialize_uncompressed(&mut proof_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let mut prepared_inputs_bytes = Vec::new();
+        let (c1, c2) =
+            split_digest(Digest::from_hex(ALLOWED_IDS_ROOT).map_err(|err| anyhow!(err))?)?;
+        let (m1, m2) = split_digest(receipt_meta)?;
+        let public_inputs = vec![c2, c1, m2, m1];
+        let prepared_inputs =
+            ark_Groth16::<Bn254>::prepare_inputs(&public_key_verification, &public_inputs)
+                .map_err(|err| anyhow!(err))?;
+        prepared_inputs
+            .serialize_uncompressed(&mut prepared_inputs_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        Ok(Self {
+            pvk: pvk_bytes,
+            proof: proof_bytes,
+            prepared_inputs: prepared_inputs_bytes,
+        })
+    }
+
+    /// Creates a Groth16 instance from the raw material generated by Circom/SnarkJS
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn from_raw(
+        raw_vk: RawVKey,
+        raw_proof: RawProof,
+        raw_public: RawPublic,
+    ) -> Result<Self, Error> {
+        let mut pvk_bytes = Vec::new();
+        let public_key_verification = raw_vk.pvk()?;
+        public_key_verification
+            .serialize_uncompressed(&mut pvk_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let groth16_seal: Groth16Seal = raw_proof.try_into()?;
+        let proof = Proof::<Bn254> {
+            a: g1_from_bytes(&groth16_seal.a)?,
+            b: g2_from_bytes(&groth16_seal.b)?,
+            c: g1_from_bytes(&groth16_seal.c)?,
+        };
+        let mut proof_bytes = Vec::new();
+        proof
+            .serialize_uncompressed(&mut proof_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        let public_inputs = raw_public.public_inputs()?;
+        let mut prepared_inputs_bytes = Vec::new();
+        let prepared_inputs =
+            ark_Groth16::<Bn254>::prepare_inputs(&public_key_verification, &public_inputs)
+                .map_err(|err| anyhow!(err))?;
+        prepared_inputs
+            .serialize_uncompressed(&mut prepared_inputs_bytes)
+            .map_err(|err| anyhow!(err))?;
+
+        Ok(Self {
+            pvk: pvk_bytes,
+            proof: proof_bytes,
+            prepared_inputs: prepared_inputs_bytes,
+        })
+    }
+
+    /// Computes the prepared verifying key used by Bonsai
+    #[cfg(not(target_os = "zkvm"))]
+    pub fn pvk() -> Result<PreparedVerifyingKey<Bn254>, Error> {
```

Comment:
This feels like an implementation detail.
```suggestion
    fn pvk() -> Result<PreparedVerifyingKey<Bn254>, Error> {
```